### PR TITLE
fix bug: the libcgroup tool incorrectly parses the cgconfig.conf file

### PR DESCRIPTION
**Pull request by lpyu001**
Issue Background:
1.Using the libcgroup tool to perform mount operations for cgroups.
2.The cgconfig.conf file is configured as follows:

mount {
cpu = /cg_sys;
cpuacct = /cg_sys;
}

3.Execute the command system restart cgconfig.
4.lssubsys -a :
cpuset
cpu,cpuacct
blkio
memory
devices
freezer
net_cls,net_prio
5.It fails to mount both cpu and cpuacct at the same time。

Issue Analysis:
1.Directly using the mount command to mount cpu and cpuacct can mount successfully.
mount -t cgroup -o cpu,cpuacct cgroup /cg_sys
2.Using strace to trace the libcgroup execution process, it was found that only cpu was mounted during the mount system call.
libcgroup strace log:
mount("cgroup","cg_sys","cgroup",0,"cpu") = -1 EBUSY
write(1,"error:cannot mount cpu to /cg_s"...,60) = 60
mount command log:
mount("cgroup","cg_sys","cgroup",0,"cpu,cpuacct") = 0

3.After analyzing the code and adding debug logs to trace the change of the mount->name variable in the cgroup_config_mount_fs function, it was found that before and after calling the cgroup_config_adjust_mount_options function, the mount->name variable changed from "cpu,cpacct" to "cpu", so it was suspected that there was an issue in how the cgroup_config_adjust_mount_options function handled the mount->name variable.
4.Further analysis of the cgroup_config_adjust_mount_options function revealed that it only processes the first controller. After extracting the first controller, the mount->name is directly overwritten using strncpy, causing only the first subsystem to be mounted.
static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount,
						unsigned long *flags)
{
	char *save = NULL;
	char *opts = strdup(mount->name);
	char *token;
	int name_only = 1;
	char *controller= NULL;
	bool is_first_controller = true;

	*flags = 0;
	if (opts == NULL)
		return ECGFAIL;

	token = strtok_r(opts, ",", &save);
	while (token != NULL) {
		if (strncmp(token, "name=", 5) != 0) {
			name_only = 0;
			if (!controller) {
               controller = strdup(token);
               if(controller = NULL) 
			      break;	
                strncpy(mount->name, controller, sizeof(mount->name));
                mount->name[sizeof(mount->name)-1] = '\0';
         	}

Issue Resolution:
A flag is_first_controller was added to determine whether it is the first controller. If it is the first, the controller is copied to mount->name using strncpy. If it is not the first, strncat is used to append subsequent controllers one by one.

### CODE REVIEW

**Comment by kamalesh-babulal** [comment]
@lpyu001  We follow the Linux kernel's guidelines for ownership and signing of patches
https://github.com/libcgroup/libcgroup/blob/main/CONTRIBUTING.md#sign-your-work, for the next version of the patch please include your `Real name <email.id>`

**Comment by lpyu001** [comment]
Thank you for your suggestions. Based on your advice, I have made improvements and verified the patch:

1.After processing the first controller, the next iteration should be triggered, otherwise the first controller will be duplicated.
2.Adjusted the position of free(controller)."


 src/config.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/config.c b/src/config.c
index 5095a50..7ab25ec 100644
--- a/src/config.c
+++ b/src/config.c
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+	int i;
+     
+	char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices", "freezer",
+                              "hugetlb", "memory", "misc", "net_cls", "net_prio",
+                              "perf_event", "pids", "rdma", NULL };
+	    
+    	for (i = 0; ctrl_list[i]; i++) {
+		if (strncmp(ctrl, ctrl_list[i], strlen(ctrl_list[i])) == 0)
+			return 1;
+    	}	
+
+    return 0;
+}
+
 /**
  * Add necessary options for mount. Currently only 'none' option is added
  * for mounts with only 'name=xxx' and without real controller.
@@ -657,6 +673,9 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 
 				strncpy(mount->name, controller, sizeof(mount->name));
 				mount->name[sizeof(mount->name)-1] = '\0';
+				free(controller);
+				token = strtok_r(NULL, ",", &save);
+				continue;
 			}
 
 			if (strncmp(token, "nodev", strlen("nodev")) == 0)
@@ -668,6 +687,15 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 			if (strncmp(token, "nosuid", strlen("nosuid")) == 0)
 				*flags |= MS_NOSUID;
 
+			if (is_valid_controller(token)) {
+				controller = strdup(token);
+				if (controller == NULL)
+					break;
+				strncat(mount->name, ",", FILENAME_MAX - strlen(mount->name)-1);
+				strncat(mount->name, controller, FILENAME_MAX - strlen(mount->name) - 1);
+				free(controller);
+                      }
+
 		} else if (!name_only) {
 			/*
 			 * We have controller + name=, do the right thing,
@@ -679,7 +707,6 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 		token = strtok_r(NULL, ",", &save);
 	}
 
-	free(controller);
 	free(opts);
 
 	if (name_only) {
-- 
2.25.1

please review ,thanks


[root@localhost ~]# cgconfigparser -l /etc/cgconfig.conf -s 1664
[root@localhost ~]#
[root@localhost ~]# mount | grep cg_sys
cgroup on /cg_sys type cgroup (rw,relatime,cpu,cpuacct)
cgroup on /cg_sys1 type cgroup (rw,relatime,memory)

**Comment by lpyu001** [comment]
Thank you for your suggestions. Based on your advice, I have made improvements and verified the patch:

1.After processing the first controller, the next iteration should be triggered, otherwise the first controller will be duplicated.
2.Adjusted the position of free(controller)."

`src/config.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/config.c b/src/config.c
index 5095a50..7ab25ec 100644
--- a/src/config.c
+++ b/src/config.c
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+	int i;
+     
+	char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices", "freezer",
+                              "hugetlb", "memory", "misc", "net_cls", "net_prio",
+                              "perf_event", "pids", "rdma", NULL };
+	    
+    	for (i = 0; ctrl_list[i]; i++) {
+		if (strncmp(ctrl, ctrl_list[i], strlen(ctrl_list[i])) == 0)
+			return 1;
+    	}	
+
+    return 0;
+}
+
 /**
  * Add necessary options for mount. Currently only 'none' option is added
  * for mounts with only 'name=xxx' and without real controller.
@@ -657,6 +673,9 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 
 				strncpy(mount->name, controller, sizeof(mount->name));
 				mount->name[sizeof(mount->name)-1] = '\0';
+				free(controller);
+				token = strtok_r(NULL, ",", &save);
+				continue;
 			}
 
 			if (strncmp(token, "nodev", strlen("nodev")) == 0)
@@ -668,6 +687,15 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 			if (strncmp(token, "nosuid", strlen("nosuid")) == 0)
 				*flags |= MS_NOSUID;
 
+			if (is_valid_controller(token)) {
+				controller = strdup(token);
+				if (controller == NULL)
+					break;
+				strncat(mount->name, ",", FILENAME_MAX - strlen(mount->name)-1);
+				strncat(mount->name, controller, FILENAME_MAX - strlen(mount->name) - 1);
+				free(controller);
+           }
+
 		} else if (!name_only) {
 			/*
 			 * We have controller + name=, do the right thing,
@@ -679,7 +707,6 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 		token = strtok_r(NULL, ",", &save);
 	}
 
-	free(controller);
 	free(opts);
 
 	if (name_only) {
-- 
2.25.1`

please review ,thanks


`[root@localhost ~]# cgconfigparser -l /etc/cgconfig.conf -s 1664
[root@localhost ~]#
[root@localhost ~]# mount | grep cg_sys
cgroup on /cg_sys type cgroup (rw,relatime,cpu,cpuacct)
cgroup on /cg_sys1 type cgroup (rw,relatime,memory)
`

**Comment by lpyu001** [comment]
>  src/config.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/config.c b/src/config.c
index 5095a50..7ab25ec 100644
--- a/src/config.c
+++ b/src/config.c
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+	int i;
+     
+	char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices", "freezer",
+                              "hugetlb", "memory", "misc", "net_cls", "net_prio",
+                              "perf_event", "pids", "rdma", NULL };
+	    
+    	for (i = 0; ctrl_list[i]; i++) {
+		if (strncmp(ctrl, ctrl_list[i], strlen(ctrl_list[i])) == 0)
+			return 1;
+    	}	
+
+    return 0;
+}
+
 /**
  * Add necessary options for mount. Currently only 'none' option is added
  * for mounts with only 'name=xxx' and without real controller.
@@ -657,6 +673,9 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 
 				strncpy(mount->name, controller, sizeof(mount->name));
 				mount->name[sizeof(mount->name)-1] = '\0';
+				free(controller);
+				token = strtok_r(NULL, ",", &save);
+				continue;
 			}
 
 			if (strncmp(token, "nodev", strlen("nodev")) == 0)
@@ -668,6 +687,15 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 			if (strncmp(token, "nosuid", strlen("nosuid")) == 0)
 				*flags |= MS_NOSUID;
 
+			if (is_valid_controller(token)) {
+				controller = strdup(token);
+				if (controller == NULL)
+					break;
+				strncat(mount->name, ",", FILENAME_MAX - strlen(mount->name)-1);
+				strncat(mount->name, controller, FILENAME_MAX - strlen(mount->name) - 1);
+				free(controller);
+           }
+
 		} else if (!name_only) {
 			/*
 			 * We have controller + name=, do the right thing,
@@ -679,7 +707,6 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 		token = strtok_r(NULL, ",", &save);
 	}
 
-	free(controller);
 	free(opts);
 
 	if (name_only) {
-- 
2.25.1

**Comment by lpyu001** [comment]
_src/config.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/config.c b/src/config.c
index 5095a50..7ab25ec 100644
--- a/src/config.c
+++ b/src/config.c
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+	int i;
+     
+	char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices", "freezer",
+                              "hugetlb", "memory", "misc", "net_cls", "net_prio",
+                              "perf_event", "pids", "rdma", NULL };
+	    
+    	for (i = 0; ctrl_list[i]; i++) {
+		if (strncmp(ctrl, ctrl_list[i], strlen(ctrl_list[i])) == 0)
+			return 1;
+    	}	
+
+    return 0;
+}
+
 /**
  * Add necessary options for mount. Currently only 'none' option is added
  * for mounts with only 'name=xxx' and without real controller.
@@ -657,6 +673,9 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 
 				strncpy(mount->name, controller, sizeof(mount->name));
 				mount->name[sizeof(mount->name)-1] = '\0';
+				free(controller);
+				token = strtok_r(NULL, ",", &save);
+				continue;
 			}
 
 			if (strncmp(token, "nodev", strlen("nodev")) == 0)
@@ -668,6 +687,15 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 			if (strncmp(token, "nosuid", strlen("nosuid")) == 0)
 				*flags |= MS_NOSUID;
 
+			if (is_valid_controller(token)) {
+				controller = strdup(token);
+				if (controller == NULL)
+					break;
+				strncat(mount->name, ",", FILENAME_MAX - strlen(mount->name)-1);
+				strncat(mount->name, controller, FILENAME_MAX - strlen(mount->name) - 1);
+				free(controller);
+           }
+
 		} else if (!name_only) {
 			/*
 			 * We have controller + name=, do the right thing,
@@ -679,7 +707,6 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 		token = strtok_r(NULL, ",", &save);
 	}
 
-	free(controller);
 	free(opts);
 
 	if (name_only) {
-- 
2.25.1_

**Comment by lpyu001** [comment]
### 
src/config.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/config.c b/src/config.c
index 5095a50..7ab25ec 100644
--- a/src/config.c
+++ b/src/config.c
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+	int i;
+     
+	char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices", "freezer",
+                              "hugetlb", "memory", "misc", "net_cls", "net_prio",
+                              "perf_event", "pids", "rdma", NULL };
+	    
+    	for (i = 0; ctrl_list[i]; i++) {
+		if (strncmp(ctrl, ctrl_list[i], strlen(ctrl_list[i])) == 0)
+			return 1;
+    	}	
+
+    return 0;
+}
+
 /**
  * Add necessary options for mount. Currently only 'none' option is added
  * for mounts with only 'name=xxx' and without real controller.
@@ -657,6 +673,9 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 
 				strncpy(mount->name, controller, sizeof(mount->name));
 				mount->name[sizeof(mount->name)-1] = '\0';
+				free(controller);
+				token = strtok_r(NULL, ",", &save);
+				continue;
 			}
 
 			if (strncmp(token, "nodev", strlen("nodev")) == 0)
@@ -668,6 +687,15 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 			if (strncmp(token, "nosuid", strlen("nosuid")) == 0)
 				*flags |= MS_NOSUID;
 
+			if (is_valid_controller(token)) {
+				controller = strdup(token);
+				if (controller == NULL)
+					break;
+				strncat(mount->name, ",", FILENAME_MAX - strlen(mount->name)-1);
+				strncat(mount->name, controller, FILENAME_MAX - strlen(mount->name) - 1);
+				free(controller);
+           }
+
 		} else if (!name_only) {
 			/*
 			 * We have controller + name=, do the right thing,
@@ -679,7 +707,6 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 		token = strtok_r(NULL, ",", &save);
 	}
 
-	free(controller);
 	free(opts);
 
 	if (name_only) {
-- 
2.25.1

**Comment by lpyu001** [comment]
***---
 src/config.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/config.c b/src/config.c
index 5095a50..7ab25ec 100644
--- a/src/config.c
+++ b/src/config.c
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+	int i;
+     
+	char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices", "freezer",
+                              "hugetlb", "memory", "misc", "net_cls", "net_prio",
+                              "perf_event", "pids", "rdma", NULL };
+	    
+    	for (i = 0; ctrl_list[i]; i++) {
+		if (strncmp(ctrl, ctrl_list[i], strlen(ctrl_list[i])) == 0)
+			return 1;
+    	}	
+
+    return 0;
+}
+
 /**
  * Add necessary options for mount. Currently only 'none' option is added
  * for mounts with only 'name=xxx' and without real controller.
@@ -657,6 +673,9 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 
 				strncpy(mount->name, controller, sizeof(mount->name));
 				mount->name[sizeof(mount->name)-1] = '\0';
+				free(controller);
+				token = strtok_r(NULL, ",", &save);
+				continue;
 			}
 
 			if (strncmp(token, "nodev", strlen("nodev")) == 0)
@@ -668,6 +687,15 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 			if (strncmp(token, "nosuid", strlen("nosuid")) == 0)
 				*flags |= MS_NOSUID;
 
+			if (is_valid_controller(token)) {
+				controller = strdup(token);
+				if (controller == NULL)
+					break;
+				strncat(mount->name, ",", FILENAME_MAX - strlen(mount->name)-1);
+				strncat(mount->name, controller, FILENAME_MAX - strlen(mount->name) - 1);
+				free(controller);
+           		}
+
 		} else if (!name_only) {
 			/*
 			 * We have controller + name=, do the right thing,
@@ -679,7 +707,6 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 		token = strtok_r(NULL, ",", &save);
 	}
 
-	free(controller);
 	free(opts);
 
 	if (name_only) {
-- 
2.25.1***

**Comment by lpyu001** [comment]
@lpyu001 Thank you so much for the patch. The patch works for most of the cases, but consider the following cgconfig.conf:

mount {
        cpu,nodev = /cg_sys;
        memory = /cg_sys_1;
        cpuacct = /cg_sys;
}
it would fail with error.

./src/tools/cgconfigparser -l /etc/cgconfig.conf 
Error: cannot mount cpu,nodev,cpuacct to /cg_sys: Invalid argument
/libcgroup/src/tools/.libs/lt-cgconfigparser; error loading /etc/cgconfig.conf: Cgroup mounting failed
Unconditionally assuming that there would be no options passed for the controller, would break.

One idea would be to check for the token is valid controller, something like:

diff --git a/src/config.c b/src/config.c
index 5095a50..835453b 100644
--- a/src/config.c
+++ b/src/config.c
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
        memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+
+       char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices", "freezer",
+                              "hugetlb", "memory", "misc", "net_cls", "net_prio",
+                              "perf_event", "pids", "rdma", NULL };
+
+       int i;
+
+       for (i = 0; ctrl_list[i]; i++) {
+               if (strncmp(ctrl, ctrl_list[i], strlen(ctrl_list[i])) == 0)
+                       return 1;
+       }
+
+       return 0;
+}
 /**
  * Add necessary options for mount. Currently only 'none' option is added
  * for mounts with only 'name=xxx' and without real controller.
@@ -657,6 +673,7 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
 
                                strncpy(mount->name, controller, sizeof(mount->name));
                                mount->name[sizeof(mount->name)-1] = '\0';
+                               free(controller);
                        }
 
                        if (strncmp(token, "nodev", strlen("nodev")) == 0)
@@ -668,6 +685,14 @@ static int cgroup_config_ajdust_mount_options(struct cg_mount_table_s *mount, un
                        if (strncmp(token, "nosuid", strlen("nosuid")) == 0)
                                *flags |= MS_NOSUID;
 
+                       if (is_valid_controller(token)) {
+                               controller = strdup(token);
+                               if (controller == NULL)
+                                       break;
+                               strncat(mount->name, ",", FILENAME_MAX - strlen(mount->name)-1);
+                               strncat(mount->name, controller, FILENAME_MAX - strlen(mount->name) - 1);
+                       }
+
                } else if (!name_only) {
                        /*
                         * We have controller + name=, do the right thing,

**Comment by lpyu001** [comment]
@kamalesh-babulal Thank you for your suggestions. Based on your advice, I have made improvements and verified the patch:

1.After processing the first controller, the next iteration should be triggered, otherwise the first controller will be duplicated.
2.Adjusted the position of free(controller)."

patch :
[0001-fix-bug-libcgroup-parse-cgconfig-error.patch](https://github.com/user-attachments/files/17974934/0001-fix-bug-libcgroup-parse-cgconfig-error.patch)

palease review,thanks

Verification result:
[root@localhost ~]# cgconfigparser -l /etc/cgconfig.conf -s 1664
[root@localhost ~]#
[root@localhost ~]# mount | grep cg_sys
cgroup on /cg_sys type cgroup (rw,relatime,cpu,cpuacct)
cgroup on /cg_sys1 type cgroup (rw,relatime,memory)

**Review by kamalesh-babulal — APPROVED** [review]

**Comment by kamalesh-babulal** [comment]
@lpyu001 Thanks for working on the v2 of the patch.

**Comment by kamalesh-babulal** [comment]
@lpyu001 Can you please the minor checkpatch review issues (https://github.com/libcgroup/libcgroup/actions/runs/12116006401/job/33779246588?pr=453)  with the patch?

**Comment by lpyu001** [comment]
> @lpyu001 Can you please the minor checkpatch review issues (https://github.com/libcgroup/libcgroup/actions/runs/12116006401/job/33779246588?pr=453) with the patch?

New code has been resubmitted，thanks

File: src/config.c [hunk 1]
```diff
@@ -629,6 +629,22 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+static int is_valid_controller(char *ctrl)
+{
+	int i;
+
+	static const char *ctrl_list[] = { "blkio", "cpu", "cpuacct", "cpuset", "devices",
+					"freezer", "hugetlb", "memory", "misc", "net_cls",
+					"net_prio", "perf_event", "pids", "rdma", NULL };
```
Comment (drakenclimber) [inline src/config.c:638]: This is a note to @kamalesh-babulal and myself.  We'll need to make sure that this list of valid controllers closely follows the upstream kernel.  If/When new controllers are added, we'll need to update this function.

**Review by drakenclimber — APPROVED** [review]
Thanks for the patch, @lpyu001.

Let's update the commit message to better capture the bug.  Perhaps something like this:

```
cgconfig: Add support for multiple controllers on the same mount path

Add support to cgconfigparser for mounting multiple controllers to the same mount point.

Given the following config:

mount {
    cpu = /cg_sys;
    cpuacct = /cg_sys;
}

cgconfigparser should mount both the cpu and cpuacct controllers to /cg_sys.  The mount should look as follows:

cgroup on /cg_sys type cgroup (rw,relatime,seclabel,cpu,cpuacct)

Without this change, the mount->name gets overwritten and the mount fails:

Error: cannot mount cpu to /cg_sys: Device or resource busy
```

**Comment by drakenclimber** [comment]
`Acked-by: Tom Hromatka <tom.hromatka@oracle.com>`

**Comment by lpyu001** [comment]
@drakenclimber the commit message has been updated

**Comment by kamalesh-babulal** [comment]
Pushed into main in commit https://github.com/libcgroup/libcgroup/commit/dcce65bde49dd86046ffd87091cfd17b2af81be5. @lpyu001 Thank you for working on the suggestions. @drakenclimber Thank you for the review.

File: src/config.c (hunk 1 above)
Comment (kamalesh-babulal) [reply src/config.c:638]: @drakenclimber Agreed, we need to keep an eye on new controllers.

**Pull request closed without merging**